File: python/sol/comm_X.py
# -*- coding: utf-8 -*-
"""
comm_X.py
"""

from mpi4py import MPI
from numba import jit, prange

# X境界の電圧を共有する
def share(Parm, V, SendBuf_x, RecvBuf_x,
    iMin, iMax, jMin, jMax, kMin, kMax, Ni, Nj, Nk, N0):

    Npx = Parm['Npx']
    Npy = Parm['Npy']
    Npz = Parm['Npz']
    Ipx = Parm['Ipx']
    Ipy = Parm['Ipy']
    Ipz = Parm['Ipz']

    bx = [Ipx > 0, Ipx < Npx - 1]
    px = [Ipx - 1, Ipx + 1]
    isend = [iMin + 1, iMax - 1]
    irecv = [iMin - 1, iMax + 1]

    for side in range(2):
        if bx[side]:
            # V to buffer
            i = isend[side]
            _V_to_buffer(i, V, SendBuf_x, jMin, jMax, kMin, kMax, Ni, Nj, Nk, N0)

            # MPI
            ipx = px[side]
            dst = (ipx * Npy * Npz) + (Ipy * Npz) + Ipz
            MPI.COMM_WORLD.Sendrecv(SendBuf_x, dst, recvbuf=RecvBuf_x, source=dst)

            # from buffer to V
            i = irecv[side]
            _buffer_to_V(i, V, RecvBuf_x, jMin, jMax, kMin, kMax, Ni, Nj, Nk, N0)

# (private) from V to buffer
@jit(cache=True, nopython=True, nogil=True, parallel=True)
def _V_to_buffer(i, V, buf, jMin, jMax, kMin, kMax, Ni, Nj, Nk, N0):

    for j in prange(jMin, jMax + 1):
        for k in range(kMin, kMax + 1):
            n = (Ni * i) + (Nj * j) + (Nk * k) + N0
            m = (j - jMin) * (kMax - kMin + 1) + (k - kMin)
            buf[m] = V[n]

# (private) from buffer to V
@jit(cache=True, nopython=True, nogil=True, parallel=True)
def _buffer_to_V(i, V, buf, jMin, jMax, kMin, kMax, Ni, Nj, Nk, N0):

    for j in prange(jMin, jMax + 1):
        for k in range(kMin, kMax + 1):
            n = (Ni * i) + (Nj * j) + (Nk * k) + N0
            m = (j - jMin) * (kMax - kMin + 1) + (k - kMin)
            V[n] = buf[m]

File: mpi/comm_X.c
/*
comm_X.c
*/

#ifdef _MPI
#include <mpi.h>
#endif

#include "ost.h"

// V to buffer
static void v_to_buffer(int i, real_t v[], real_t buf[], int j0, int j1, int k0, int k1)
{
	for (int j = j0; j <= j1; j++) {
	for (int k = k0; k <= k1; k++) {
		const int64_t m = (j - j0) * (k1 - k0 + 1) + (k - k0);
		buf[m] = v[D(i, j, k)];
	}
	}
}


// buffer to V
static void buffer_to_v(int i, real_t v[], real_t buf[], int j0, int j1, int k0, int k1)
{
	for (int j = j0; j <= j1; j++) {
	for (int k = k0; k <= k1; k++) {
		const int64_t m = (j - j0) * (k1 - k0 + 1) + (k - k0);
		v[D(i, j, k)] = buf[m];
	}
	}
}


void comm_X(void)
{
#ifdef _MPI
	MPI_Status status;
	const int tag = 0;
	int bx[] = {Ipx > 0, Ipx < Npx - 1};
	int px[] = {Ipx - 1, Ipx + 1};
	int isend[] = {iMin + 1, iMax - 1};
	int irecv[] = {iMin - 1, iMax + 1};
	int i;

	for (int side = 0; side < 2; side++) {
		if (bx[side]) {
			// V to buffer
			i = isend[side];
			v_to_buffer(i, V, SendBuf_x, jMin, jMax, kMin, kMax);

			// MPI
			const int ipx = px[side];
			const int dst = (ipx * Npy * Npz) + (Ipy * Npz) + Ipz;
			const int count = (jMax - jMin + 1) * (kMax - kMin + 1);
			MPI_Sendrecv(SendBuf_x, count, MPI_REAL_T, dst, tag,
			             RecvBuf_x, count, MPI_REAL_T, dst, tag, MPI_COMM_WORLD, &status);

			// buffer to V
			i = irecv[side];
			buffer_to_v(i, V, RecvBuf_x, jMin, jMax, kMin, kMax);
		}
	}
#endif
}
